### ISSUE TITLE
ng update on windows fails when cli version wants to preinstall | loader.js | Cannot find module

### ISSUE THREAD

**Issue by florian-kittel** [angular/angular-cli#18891]
# 🐞 Bug report

### Command (mark with an `x`)

- [ x ] update

### Is this a regression?

 Yes, the previous version in which this bug was not present was: 6, 7, partially in 8 (depend on the cli version)

### Description

A user directory on windows is 'C:\Users\Firstname Lastname\...' - what mean it includes a whitespace between the first and last name. When the given project has not the latest cli version installed, a temporary install happen of the lastes cli version to use it for ng update. Then the following error happen (I holde it generic by using "Firstname Lastname":

```
$ ng update
The installed local Angular CLI version is older than the latest stable version.
Installing a temporary version to perform the update.
Installing packages for tooling via npm.
Installed packages for tooling via npm.
internal/modules/cjs/loader.js:1083
  throw err;
  ^

Error: Cannot find module 'C:\Users\Firstname'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:1080:15)
    at Function.Module._load (internal/modules/cjs/loader.js:923:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

## 🔬 Minimal Reproduction

Have a windows system where the user name uses a whitespace (like "Firstname Lastname" or "John Doe") and an Angular project that has not the latest cli installed. Run `ng update`.

## 🔥 Exception or Error
<pre><code>
$ ng update
The installed local Angular CLI version is older than the latest stable version.
Installing a temporary version to perform the update.
Installing packages for tooling via npm.
Installed packages for tooling via npm.
internal/modules/cjs/loader.js:1083
  throw err;
  ^

Error: Cannot find module 'C:\Users\Firstname'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:1080:15)
    at Function.Module._load (internal/modules/cjs/loader.js:923:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}</code></pre>

## 🌍 Your Environment
<pre><code>
Angular CLI: 10.1.1
Node: 14.8.0
OS: win32 x64

Angular: 10.1.1
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.1001.1
@angular-devkit/build-angular      0.1001.1
@angular-devkit/build-ng-packagr   0.1001.1
@angular-devkit/build-optimizer    0.1001.1
@angular-devkit/build-webpack      0.1001.1
@angular-devkit/core               10.1.1
@angular-devkit/schematics         10.1.1
@ngtools/webpack                   10.1.1
@schematics/angular                10.1.1
@schematics/update                 0.1001.1
ng-packagr                         10.1.0
rxjs                               6.6.3
typescript                         4.0.2
webpack                            4.44.1
</code></pre>

**Anything else relevant?**
To use `ng update` it is possible to update the local pacakge to latest version by increasing it in package.json by hand and do an `npm install`. Then with the lastest version `ng update` always works. It is the problem with the mechanism to preinstall the lastes version on `ng update` command. 

I also tried to route the global node_modules to a different folder. That works for all global packages, but not for that command. It will always use the users directory.

**Comment by alan-agius4** [comment]
Related to https://github.com/angular/angular-cli/issues/17870

**Comment by Richardo86** [comment]
> # 🐞 Bug report
> ### Command (mark with an `x`)
> * [ x ] update
> 
> ### Is this a regression?
> Yes, the previous version in which this bug was not present was: 6, 7, partially in 8 (depend on the cli version)
> ### Description
> A user directory on windows is 'C:\Users\Firstname Lastname...' - what mean it includes a whitespace between the first and last name. When the given project has not the latest cli version installed, a temporary install happen of the lastes cli version to use it for ng update. Then the following error happen (I holde it generic by using "Firstname Lastname":
> 
> ```
> $ ng update
> The installed local Angular CLI version is older than the latest stable version.
> Installing a temporary version to perform the update.
> Installing packages for tooling via npm.
> Installed packages for tooling via npm.
> internal/modules/cjs/loader.js:1083
>   throw err;
>   ^
> 
> Error: Cannot find module 'C:\Users\Firstname'
>     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:1080:15)
>     at Function.Module._load (internal/modules/cjs/loader.js:923:27)
>     at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
>     at internal/main/run_main_module.js:17:47 {
>   code: 'MODULE_NOT_FOUND',
>   requireStack: []
> }
> ```
> 
> ## 🔬 Minimal Reproduction
> Have a windows system where the user name uses a whitespace (like "Firstname Lastname" or "John Doe") and an Angular project that has not the latest cli installed. Run `ng update`.
> 
> ## 🔥 Exception or Error
> ```
> 
> $ ng update
> The installed local Angular CLI version is older than the latest stable version.
> Installing a temporary version to perform the update.
> Installing packages for tooling via npm.
> Installed packages for tooling via npm.
> internal/modules/cjs/loader.js:1083
>   throw err;
>   ^
> 
> Error: Cannot find module 'C:\Users\Firstname'
>     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:1080:15)
>     at Function.Module._load (internal/modules/cjs/loader.js:923:27)
>     at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
>     at internal/main/run_main_module.js:17:47 {
>   code: 'MODULE_NOT_FOUND',
>   requireStack: []
> }
> ```
> 
> ## 🌍 Your Environment
> ```
> 
> Angular CLI: 10.1.1
> Node: 14.8.0
> OS: win32 x64
> 
> Angular: 10.1.1
> ... animations, cli, common, compiler, compiler-cli, core, forms
> ... language-service, platform-browser, platform-browser-dynamic
> ... router
> Ivy Workspace: Yes
> 
> Package                            Version
> ------------------------------------------------------------
> @angular-devkit/architect          0.1001.1
> @angular-devkit/build-angular      0.1001.1
> @angular-devkit/build-ng-packagr   0.1001.1
> @angular-devkit/build-optimizer    0.1001.1
> @angular-devkit/build-webpack      0.1001.1
> @angular-devkit/core               10.1.1
> @angular-devkit/schematics         10.1.1
> @ngtools/webpack                   10.1.1
> @schematics/angular                10.1.1
> @schematics/update                 0.1001.1
> ng-packagr                         10.1.0
> rxjs                               6.6.3
> typescript                         4.0.2
> webpack                            4.44.1
> ```
> 
> **Anything else relevant?**
> To use `ng update` it is possible to update the local pacakge to latest version by increasing it in package.json by hand and do an `npm install`. Then with the lastest version `ng update` always works. It is the problem with the mechanism to preinstall the lastes version on `ng update` command.
> 
> I also tried to route the global node_modules to a different folder. That works for all global packages, but not for that command. It will always use the users directory.

**Comment by Richardo86** [comment]
Angular CLI: 10.1.1
Node: 14.8.0
OS: win32 x64

Angular: 10.1.1
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.1001.1
@angular-devkit/build-angular      0.1001.1
@angular-devkit/build-ng-packagr   0.1001.1
@angular-devkit/build-optimizer    0.1001.1
@angular-devkit/build-webpack      0.1001.1
@angular-devkit/core               10.1.1
@angular-devkit/schematics         10.1.1
@ngtools/webpack                   10.1.1
@schematics/angular                10.1.1
@schematics/update                 0.1001.1
ng-packagr                         10.1.0
rxjs                               6.6.3
typescript                         4.0.2
webpack                            4.44.1

**Comment by Richardo86** [comment]
$ ng update
The installed local Angular CLI version is older than the latest stable version.
Installing a temporary version to perform the update.
Installing packages for tooling via npm.
Installed packages for tooling via npm.
internal/modules/cjs/loader.js:1083
  throw err;
  ^

Error: Cannot find module 'C:\Users\Firstname'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:1080:15)
    at Function.Module._load (internal/modules/cjs/loader.js:923:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

**Comment by inidaname** [comment]
Hello @alan-agius4 
I would love to take a look at this

**Comment by alan-agius4** [comment]
Go for it @inidaname.

**Comment by inidaname** [comment]
Hello @alan-agius4 

Thank you, could you point me to a guide of what is expected, I see you suggested on #17870 to implement path escape could you elaborate on that.

Thank you

**Comment by inidaname** [comment]
@alan-agius4
going through a similar issue on another project one of the suggested fixes mentioned on https://github.com/nodejs/node/issues/7367#issuecomment-229721296 is to add an option of ` { shell: true }` is this a fix that would be welcome, due to other concern mentioned in the same thread https://github.com/nodejs/node/issues/7367#issuecomment-238594729

**Comment by alan-agius4** [comment]
Hi @inidaname, we use `shell: true` option and hence to why we need to escape the path. I think wrapped them in quotes should be enough, but this definitely needs some testing on a Windows machine.

https://github.com/angular/angular-cli/blob/57640bebfd0de417456db6c5d3cead45d528c055/packages/angular/cli/utilities/install-package.ts#L211

**Comment by Sarah-JJ** [comment]
> 
> 
> Hi @inidaname, we use `shell: true` option
> 
> https://github.com/angular/angular-cli/blob/57640bebfd0de417456db6c5d3cead45d528c055/packages/angular/cli/utilities/install-package.ts#L211
> and hence to why we need to escape the path.
> 
> As mentioned in the issue linked, I think wrapped the `binPath`
> 
> https://github.com/angular/angular-cli/blob/57640bebfd0de417456db6c5d3cead45d528c055/packages/angular/cli/utilities/install-package.ts#L207
> in quotes should be enough, but this definitely needs some testing on a Windows machine.

In the project folder in \node_modules\@angular\cli\utilities\install-package.js, I replaced 
`const argv = [`'${binPath}'`, ...args];`
with 
`const argv = [binPath, ...args];`
And `ng update` started working again

**Comment by alan-agius4** [comment]
@Sarah-JJ, please update the global CLI to the latest version as the mentioned issue has already been resolved in 12.0.3.

**Comment by UserPieter** [comment]
@Sarah-JJ this solution worked for me but there was a small difference in the line of code you refer to. 

it said: `const argv = [`'${binPath}'`, ...args];
`
i removed the inner quotes and then I could update.

@alan-agius4 : that solution didnt work for me. i was already using version 12.0.3 but I did also install it globally. it didnt change anything, the error was still there.
